src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs

```diff
@@ -7490,7 +7530,20 @@ internal Task WriteString(string s, int length, int offset, TdsParserStateObject
 
         private static void CopyCharsToBytes(char[] source, int sourceOffset, byte[] dest, int destOffset, int charLength)
         {
-            Buffer.BlockCopy(source, sourceOffset, dest, destOffset, charLength * ADP.CharSize);
+            if (!BitConverter.IsLittleEndian)
+            {
+                int desti = 0;
+                for(int srci = 0; srci < charLength; srci++)
+                {
+                    dest[desti + destOffset] = (byte)(source[srci + sourceOffset]);
+                    dest[desti + destOffset+1] = (byte)(source[srci + sourceOffset] >> 8);
+                    desti += 2;
+                }
```

Comment:
why not `BinaryPrimitives.ReverseEndianness(ushort)` ?

Reply:
we have to explicity cast it to (byte) and we lose 1 byte so we would have to shift it again to incorporate in the dest byte array which these two statements already do

Reply:
Maybe it will be slightly faster if you perform one read operation rather than two (e.g. `char c = source[srci + sourceOffset]` than use `c`). I have no idea whether the compiler does this optimization automatically.

Reply:
You could also do `BinaryPrimitives.WriteInt16LittleEndian(dest.AsSpan(desti + destOffset), (short)source[srci + sourceOffset]);` in the loop.

Since the system is not little endian, writing in little endian effectively reverses the endianness. I think @Wraith2 meant something along these lines.